### PR TITLE
Adding Per Message Compression Extensions from RFC-7692.

### PR DESCRIPTION
**Pull request by jhoughjr**
My first PR is PMCE support for WebSocket server and client.

There is an accompanying PR going to Vapor to support the changes to the WebSocketUpgrader to enable conditional use of PMCE.

I've tried to comment everything as descriptive as possible but without knowledge of the RFC, the code may not make obvious sense.

The changes are mostly encapsulated in the new PMCE class and its members.
Websocket's main changes are an optional PMCE which is Sendable and some optional delegation to it if configured and enabled for use in the send and handle methods.

WebSocketClient.Configuration also has an optional PMCE so the client can configure its side.
The public API is mainly through converting DeflateConfigs <> HTTPHeaders.
ie if you want to serve pace web sockets, create the desire DeflateConfig and grab its headers.
Pass them out of shouldUpgrade. As a client, do the same and pass them into the connect override with the ClientConfiguration.

I have some test apps and examples in another repo and am contiuning to write documentation.
I have tested client against a node.js , a go and a vapor/websocketkit (this) implementation of PMCE servers.
I have tested the vapor/websocket-kit server against the vapor/websocket-kit client, and a web browser client, all of which works, but I'm sure its not perfect yet.

I believe MahdiBM has some experience with this issue as well as adamfowler and may be useful ion review.
I can answer any questions in discord.

### CODE REVIEW

**Review by 0xTim — COMMENTED** [review]
Some quick initial notes on this - the PR needs tidying up to start with, there's extra whitespace added, commented out code, removed tests etc

The main issue is adding a dependency on something that is not yet 1.0.0 - this is a blocker as we can't guarantee API breaking changes. This will need to be tagged a 1.0.0 before we can consider the PR (or inline the code we need)

**Comment by koraykoska** [comment]
@0xTim You are not serious, are you? As much as I like Vapor and use it every day, websocket-kit has had **multiple** breaking changes in the last years. Most recently, one of the latest releases (with the Sendable stuff) broke literally every single app I had using websocket-kit directly, within a minor version. And I can prove it if you need to.

There is literally no QA going on in any of the Vapor repositories. I trust @adam-fowler with compress-nio in a pre-1.0 version 10000% more than any of the Vapor repositories, especially websocket-kit, to not introduce any breaking changes.

Copying over code will very surely introduce another weak point in a barely maintained repository. I and surely many others would much rather depend on a maintained version of code doing a crucial part of what's being implemented in this PR.

If you really insist, why not simply version-lock it? Or reduce to bug-fix releases only? And then manually update the dependency if it makes (probably non-breaking) minor bumps.

This PR is very good and feature-complete. It even takes care of edge cases covered in the RFC. It introduces a long-awaited feature that is a crucial part of any real websocket library in any language. Vapor being the de-facto library used for Swift backends, you are the only one of any languages I know without websocket compression support. I would put priority on getting this merged ASAP.

A thorough code review will follow.

**Comment by Joannis** [comment]
Also want to mention this exists
image

File: Sources/WebSocketKit/WebSocket.swift [hunk 1]
```diff
@@ -56,9 +88,11 @@ public final class WebSocket: Sendable {
         self._closeCode = .init(nil)
         self.frameSequence = .init(nil)
         self._pingInterval = .init(nil)
+        self.pmce = nil
     }
 
     @preconcurrency public func onText(_ callback: @Sendable @escaping (WebSocket, String) -> ()) {
+        print("\(self.onTextCallback._eventLoop)")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocket.swift:95]: Print statements must be removed

File: Sources/WebSocketKit/WebSocket.swift [hunk 2]
```diff
@@ -164,21 +200,37 @@ public final class WebSocket: Sendable {
...
-        let frame = WebSocketFrame(
-            fin: fin,
-            opcode: opcode,
-            maskKey: self.makeMaskKey(),
-            data: data
-        )
-        self.channel.writeAndFlush(frame, promise: promise)
-    }
+        //using compression
+        if pmce != nil && pmce!.enabled {
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocket.swift:204]: Please `if let` unwrap this, instead of force unwrapping

File: Sources/WebSocketKit/WebSocket.swift [hunk 3]
```diff
@@ -272,22 +327,65 @@ public final class WebSocket: Sendable {
...
+                        self.frameSequence.withLockedValue { currentFrameSequence in
+                            var frameSequence = currentFrameSequence ?? .init(type: frame.opcode)
+                            // append this frame and update the sequence
+                            frameSequence.append(newFrame)
+                            currentFrameSequence = frameSequence
+                        }
+                    }
+                }
+                catch {
+                    print("websocket-kit: \(error)")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocket.swift:361]: Get rid of this print statement, and add some error handling. Of course, you can keep logging in here through `Logger`

File: Sources/WebSocketKit/WebSocket.swift [hunk 4]
```diff
@@ -272,22 +327,65 @@ public final class WebSocket: Sendable {
...
+                }
             }
+            else if frame.rsv1 && pmce == nil {
+                if pmce?.logging ?? false {
+                    print("websocket-kit: PMCE:  received compressed frame without PMCE configured! Closing per RFC-7692. You could have a configuration issue.")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocket.swift:366]: Use swift-log

File: Sources/WebSocketKit/WebSocketClient.swift [hunk 5]
```diff
@@ -34,7 +35,23 @@ public final class WebSocketClient: Sendable {
...
+        // new init to support passing in PMCE.PMCEConfig
+        public init(deflateConfig:PMCE.DeflateConfig?,
+                    tlsConfiguration:TLSConfiguration? = nil,
+                    maxFrameSize:Int = 1 << 14) {
+            
+            self.tlsConfiguration = tlsConfiguration
+            self.maxFrameSize = maxFrameSize
+            self.minNonFinalFragmentSize = 0
+            self.maxAccumulatedFrameCount = Int.max
+            self.maxAccumulatedFrameSize = Int.max
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketClient.swift:51]: I don't think the max frame sizes should _ever_ be set to `.max` implicitly.

File: Sources/WebSocketKit/WebSocketClient.swift [hunk 6]
```diff
@@ -124,21 +143,43 @@ public final class WebSocketClient: Sendable {
                         upgradeRequestHeaders.add(contentsOf: proxyHeaders)
                     }
                 }
+                print("websocket-client: config is \(self.configuration)")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketClient.swift:146]: This print statement has to be removed

File: Sources/WebSocketKit/WebSocketClient.swift [hunk 7]
```diff
@@ -44,9 +61,11 @@ public final class WebSocketClient: Sendable {
             self.minNonFinalFragmentSize = 0
             self.maxAccumulatedFrameCount = Int.max
             self.maxAccumulatedFrameSize = Int.max
+            self.deflateConfig = nil
         }
     }
 
+    // might ned to mirror above init with omce alsoi
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketClient.swift:68]: What do you mean here?

File: Sources/WebSocketKit/WebSocketClient.swift [hunk 8]
```diff
@@ -124,21 +143,43 @@ public final class WebSocketClient: Sendable {
                         upgradeRequestHeaders.add(contentsOf: proxyHeaders)
                     }
                 }
+                print("websocket-client: config is \(self.configuration)")
 
+                if upgradeRequestHeaders.contains(name: "sec-websocket-extensions") {
+                
+                    print("websocket-client: upgrade request headers are \(upgradeRequestHeaders)")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketClient.swift:150]: This is a great use case for logging again. Though I'm not sure we'll need request headers in our logs. What's the reasoning here?

File: Sources/WebSocketKit/WebSocketHandler.swift [hunk 9]
```diff
@@ -14,17 +14,30 @@ extension WebSocket {
...
         public var maxAccumulatedFrameSize: Int
-
+        
+        /// Enables PMCE if present. Defaults to 'nil'
+        /// See`PMCE` for details.
+        public var deflateConfig: PMCE.DeflateConfig? = nil
+        
+        public init(withDeflateConfig deflateConfig:PMCE.DeflateConfig) {
+            self.minNonFinalFragmentSize = 0
+            self.maxAccumulatedFrameCount = Int.max
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketHandler.swift:24]: Again, we don't want to implicitly set frame sizes to accumulate completely. This opens us up to denial of service aggacks

File: Sources/WebSocketKit/WebSocketHandler.swift [hunk 10]
```diff
@@ -90,7 +108,17 @@ extension WebSocket {
         with config: Configuration,
         onUpgrade: @Sendable @escaping (WebSocket) -> ()
     ) -> EventLoopFuture<Void> {
-        let webSocket = WebSocket(channel: channel, type: type)
+        
+        let webSocket =  config.deflateConfig != nil ?
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketHandler.swift:112]: You can simplify this by defining `let webSocket: WebSocket` first:

```swift
let webSocket: WebSocket

if let deflateConfig = config.deflateConfig {
  webSocket = WebSocket(channel: channel,
    type: type,
    pmce: PMCE(config:config.deflateConfig!,
      channel: channel,
      socketType: type
    )
  )
} else {
  webSocket = WebSocket(channel: channel, type: type)
}
```

File: Sources/WebSocketKit/WebSocketHandler.swift [hunk 11]
```diff
@@ -90,7 +108,17 @@ extension WebSocket {
...
+        WebSocket(channel: channel,
+                  type: type,
+                  pmce: PMCE(config:config.deflateConfig!,
+                             channel: channel,
+                             socketType: type)) :
+        WebSocket(channel: channel,
+                  type: type)
+        
+        print("websocket-kit: WebSocket.configure(on:...) created websocket \(webSocket) \nof \(type) \nwith \(config)")
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocketHandler.swift:121]: I think logging whether deflate was enabled _could_ be useful in an info message. But not a print

Comment (Joannis) [reply Sources/WebSocketKit/WebSocketHandler.swift:121]: Also consider using `#function` to get the current function this log was produced in

File: Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift [hunk 12]
```diff
@@ -48,7 +48,8 @@ final class AsyncWebSocketKitTests: XCTestCase {
     
     func testBadURLInWebsocketConnect() async throws {
         do {
-            try await WebSocket.connect(to: "%w", on: self.elg, onUpgrade: { _ async in })
+            // %w seems to now get to NIO and it attempts to connect to localhost:80 ... empty string makes the test pass
+            try await WebSocket.connect(to: "", on: self.elg, onUpgrade: { _ async in })
```
Comment (Joannis) [inline Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift:52]: I don't think this test should be adjusted lightheartedly. Why does it pass through to NIO?

File: Tests/WebSocketKitTests/PMCETests.swift [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+class PMCEConfigTests:XCTestCase {
+    typealias Config = PMCE.DeflateConfig
+   
+    
+}
+
+// design tests
+class PMCETests:XCTestCase {
+
```
Comment (Joannis) [inline Tests/WebSocketKitTests/PMCETests.swift:19]: This really has to be filled in

File: Tests/WebSocketKitTests/WebSocketKitTests.swift [hunk 14]
```diff
@@ -6,11 +6,18 @@ import NIOHTTP1
 import NIOSSL
 import NIOWebSocket
 @testable import WebSocketKit
+import CompressNIO
 
 final class WebSocketKitTests: XCTestCase {
+    var elg: EventLoopGroup!
```
Comment (Joannis) [inline Tests/WebSocketKitTests/WebSocketKitTests.swift:12]: Do we need this here? It's not created in `setUp` below. We seeem to have 2 `setUp` methods now. Can we make it one?

File: Tests/WebSocketKitTests/WebSocketKitTests.swift [hunk 15]
```diff
@@ -6,11 +6,18 @@ import NIOHTTP1
...
+
     override func setUp() async throws {
         fflush(stdout)
     }
+    
+    override func tearDown() {
+        try! self.elg.syncShutdownGracefully()
```
Comment (Joannis) [inline Tests/WebSocketKitTests/WebSocketKitTests.swift:19]: Not a fan of force unwrapped optionals. And `try!`. You can make `tearDown()` throwing.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift [hunk 16]
```diff
@@ -279,42 +287,6 @@ final class WebSocketKitTests: XCTestCase {
         try server.close(mode: .all).wait()
     }
 
-    func testLocally() throws {
```
Comment (Joannis) [inline Tests/WebSocketKitTests/WebSocketKitTests.swift:282]: Why was this removed?

File: Tests/WebSocketKitTests/WebSocketKitTests.swift [hunk 17]
```diff
@@ -479,7 +451,9 @@ final class WebSocketKitTests: XCTestCase {
     }
     
     func testBadURLInWebsocketConnect() async throws {
-        XCTAssertThrowsError(try WebSocket.connect(to: "%w", on: self.elg, onUpgrade: { _ in }).wait()) {
+        // %w seems to now get to NIO and it attempts to connect to localhost:80 ... empty string makes the test pass
+        XCTAssertThrowsError(try WebSocket.connect(to: "", on: self.elg, onUpgrade: { _ in }).wait()) {
```
Comment (Joannis) [inline Tests/WebSocketKitTests/WebSocketKitTests.swift:455]: Again, let's make sure we don't change tests just to get things passing.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift [hunk 18]
```diff
@@ -551,14 +528,126 @@ final class WebSocketKitTests: XCTestCase {
         try client.syncShutdown()
     }
 
-    var elg: EventLoopGroup!
+    
+    
+    // compression usage tests
```
Comment (Joannis) [inline Tests/WebSocketKitTests/WebSocketKitTests.swift:533]: Let's remove this old (commented) testing code

File: Sources/WebSocketKit/WebSocket.swift [hunk 19]
```diff
@@ -4,21 +4,24 @@ import NIOHTTP1
 import NIOSSL
 import Foundation
 import NIOFoundationCompat
+import CompressNIO
 import NIOConcurrencyHelpers
 
 public final class WebSocket: Sendable {
-    enum PeerType: Sendable {
+    
+    public enum PeerType: Sendable {
```
Comment (Joannis) [inline Sources/WebSocketKit/WebSocket.swift:12]: This doesn't need to be public

**Review by Joannis — CHANGES_REQUESTED** [review]
Code quality should really be improved, especially logging. And tests need to be added. I think this is a great start for a great addition. I'll have to read up on PMCE, but I think that the implementation is definitely a great addition and pretty well implemented.

File: Sources/WebSocketKit/WebSocket.swift (hunk 4 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift:366]: That would mean adding a dep to websocket kit for it I assume. I had looked into it possibly but thought it kinda out of scope and thought if ppl wanted logger there they would have one by now and maybe there was a reason websocket had no logger.

File: Sources/WebSocketKit/WebSocketClient.swift (hunk 5 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocketClient.swift:51]: that might be a good catch. Not sure where I saw that from, thought I pulled it from elsewhere in the codebase.
Do you know what default should be?

File: Sources/WebSocketKit/WebSocketHandler.swift (hunk 11 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocketHandler.swift:121]: these are great suggestions, I had suffered a lack of the logger for a while.

File: Tests/WebSocketKitTests/PMCETests.swift (hunk 13 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/PMCETests.swift:19]: I might remove it as I HAD tests that sendable refactor made irrelevant. They were all just commented out but I nuked them as the refactor changes all the add to be invalid tests really. Im not sure of the value the would have really as just having tests isnt the same as having useful tests that actually tell you something. 
I will ponder it and try to add something at least.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 14 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:12]: sigh. yah that might have  happened when I merged we probably dont want two setup methods.

File: Sources/WebSocketKit/WebSocket.swift (hunk 4 above)
Comment (koraykoska) [reply Sources/WebSocketKit/WebSocket.swift:366]: How is logging handled right now in websocket-kit? Simply no logging or with prints? If it's no logging I would keep it that way and make a separate PR eventually adding proper logging and the dependency.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 15 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:19]: I think those were there originally I dont recall writing them.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 16 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:282]: I had thought it was something I had added to test locally. It hangs forever the only test that didnt ever pass .

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 17 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:455]: The test is no longer valid and thus did not pass when it in fact should have passed because it wasnt testing what it said it was testing. it expects %w to return nil not a URL if I recall. As a QA engineer I take tests pretty seriously.

File: Sources/WebSocketKit/WebSocket.swift (hunk 19 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift:12]: I think it's actually needed to be public but I'll check once more. Seems it wa a couple days ago that I changed it so might not have the same api.

File: Sources/WebSocketKit/WebSocket.swift (hunk 1 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift:95]: PMCE RFC is somewhat obtuse. I'd be very surprised if it were perfectly implemented currently but it has worked against 4 different servers in my lab tests. These are all good suggestions and ive addressed most of them I think.

File: Sources/WebSocketKit/WebSocket.swift (hunk 4 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift:366]: It was no logging. I asked Tim and he agreed it shold be added. I can't make another fork though so I just did it on this PR. It was kind of a pain point to debug it without logging so I added what I could after asking around for a while. But logging has been added via swift log to websocket and pmce classes with a boxed bool to enable with a defeat of false. that way you can turn it on for real time issues without needed a fork to add it.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 15 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:19]: Ive removed the !.s I can find. Not sure what to do about the ones some other maintainer left in there. 
I can try if its important but feel it could unravel things maybe.

**Comment by Joannis** [comment]
As per [this comment](https://github.com/vapor/websocket-kit/pull/142#discussion_r1254551588) you could consider integration testing with other implementations

File: Sources/WebSocketKit/WebSocket.swift (hunk 4 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift]: Logging is added now on main.

File: Sources/WebSocketKit/WebSocketClient.swift (hunk 5 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocketClient.swift:52]: I feel like I copied it from elsewhere, but could have pulled it from somehwere... I just looked. I got it from the pre existing TLS  init I was basing adding pmce arg on ```public init(
            tlsConfiguration: TLSConfiguration? = nil,
            maxFrameSize: Int = 1 << 14
        ) ``` so I was using vapor as the guide as far as I could tell.
I do see ur point of the caller should set that but I see no way for them to pass it on init.
Seeing as it's all sendable now, shouldn't those be loop bound  to be mutable? That would prevent them from chasing it after the fact.
Im surprised it doesnt warn me. as it did on WebSocket having mutable state while being sendable. am confused.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 15 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:19]: I dont mind fixing them, but without understanding the test I cant say my change would still be a valid test. I did that in the cases I did change, like with the URL, but didnt want to go down the rabbit hole like a cowboy. Someone should do it I think. Since I already fixed deprecations without asking it's kinda six of one half dozen of another.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 16 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:282]: Just adding what I mentioned in discord that I removed it because it was skipped anyway and mistakenly thought had made it due to its vague name as I do that when drafting a test. Especially since not skipped it would hang forever.
Adding it back just adds a skipped taste is the takeaway I htink.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 18 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:533]: I just noticed the tests I added to websocket kit dont seem to be here ....!?

Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift]: Should be fixed now.

**Comment by jhoughjr** [comment]
The coverage seems odd to drop so much. I only removed a skipped test and added some that werent there. I don't understand that result.

File: Sources/WebSocketKit/WebSocket.swift (hunk 2 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift]: done.

File: Sources/WebSocketKit/WebSocketClient.swift (hunk 6 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocketClient.swift]: done

File: Sources/WebSocketKit/WebSocket.swift (hunk 19 above)
Comment (jhoughjr) [reply Sources/WebSocketKit/WebSocket.swift:14]: I just checked and it needs to be public because types outside Websocket need it.
internal doesnt suffice because an inter needs it. I really think this should be like surface protein for a websocket. It's a perfectly fine enum and is useful for the classes around websocket to support its use.

**Comment by adam-fowler** [comment]
> @0xTim You are not serious, are you?

@0xTim is right here to question relying on a package that hasn't hit v1.0. CompressNIO has had a number of breaking changes in the last few releases. Although many of those have been to add support needed by both this PR and the equivalent in hummingbird-websocket. 

To make things easier I will move compress-nio to v1.0 in the next week or so.

**Comment by jhoughjr** [comment]
@adam-fowler Thanks a ton. Your quick responsiveness and guidance has been instrumental in getting this done on time for me.

File: Tests/WebSocketKitTests/WebSocketKitTests.swift (hunk 17 above)
Comment (jhoughjr) [reply Tests/WebSocketKitTests/WebSocketKitTests.swift:455]: I'd add I tested this with a temporary test of URL init with both params and discerning what the test was testing, from that result and the test expectations. So I changed the setup to satisfy the expectation that existed prior.
I was surprised %w returned a URL and not nil.

**Comment by koraykoska** [comment]
Thank you @adam-fowler

**Comment by 0xTim** [comment]
Hey @jhoughjr any particular reason this was closed? With the recent changes and compression 1.0 this was getting into a good shape
